components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java

```diff
@@ -2267,6 +2267,18 @@ private String validatePKCEParameters(OAuthMessage oAuthMessage, OAuth2ClientVal
         return null;
     }
 
+    /* Check if PKCE is mandatory for the application.
+     *
+     * @param oAuthMessage          OAuthMessage
+     * @param validationResponse    client validation response.
+     * @return boolean Returns whether PKCE is mandatory or not.
+     */
+    private boolean isPKCEMandatory(OAuthMessage oAuthMessage,
+                                    OAuth2ClientValidationResponseDTO validationResponse) {
+        return validationResponse.isPkceMandatory() && !(Boolean.FALSE.equals
+                (oAuthMessage.getRequest().getAttribute(OAuthConstants.IS_PKCE_MANDATORY)));
```

Comment:
Is there any possibility that this attribute can be set from an outside request?
Instead of doing the check here we can do this in the `populateValidationResponseWithAppDetail` method and set the value to the validationResponse itself. Also instead of the sort of double negative check we can set a flag like IS_PKCE_UNSUPPORTED_FLOW and only even check for mandatory if that flag is set to false

Reply:
`Is there any possibility that this attribute can be set from an outside request?` what do you mean by outside request ?

Reply:
`we can set a flag like IS_PKCE_UNSUPPORTED_FLOW and only even check for mandatory if that flag is set to false` In most cases the flag value will be false, so we need to check the mandatory. Instead of setting a flag what do you think about the following approach ?

```
 private boolean isPKCEMandatory(OAuthAppDO appDO, OAuthMessage oAuthMessage){

        if (oAuthMessage.getRequest().getAttribute(OAuthConstants.PKCE_UNSUPPORTED_FLOW) != null &&
                oAuthMessage.getRequest().getAttribute(OAuthConstants.PKCE_UNSUPPORTED_FLOW).equals(true)) {
            return false;
        } else {
            return appDO.isPkceMandatory();
        }
    }
```

Reply:
> `Is there any possibility that this attribute can be set from an outside request?` what do you mean by outside request ?

This is part of the HTTP request right? can the attribute be set from a HTTP request coming to the server?

for the other question, what you have mentioned is actually similar to what I suggested. The only thing is to do it inside the `populateValidationResponseWithAppDetail` method and assign the value to the `validationResponse` itself as that is the place where we assign the pkce mandatory value in the `validationResponse`. Then we don't need this method (`isPKCEMandatory`)

Reply:
`can the attribute be set from a HTTP request coming to the server?` no that's not possible. Attribute can be set only from the server side.

Reply:
Addressed with https://github.com/wso2-extensions/identity-inbound-auth-oauth/pull/2112/commits/b738e13321fada46d09f7207717d7338aa3e3dd7